uspace/app/layout/layout.c

```diff
@@ -0,0 +1,178 @@
+/*
+ * Copyright (c) 2019 Matthieu Riolo
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * - Redistributions of source code must retain the above copyright
+ *   notice, this list of conditions and the following disclaimer.
+ * - Redistributions in binary form must reproduce the above copyright
+ *   notice, this list of conditions and the following disclaimer in the
+ *   documentation and/or other materials provided with the distribution.
+ * - The name of the author may not be used to endorse or promote products
+ *   derived from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
+ * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+ * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
+ * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+ * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/** @addtogroup layout
+ * @{
+ */
+/**
+ * @file
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <str.h>
+#include <str_error.h>
+#include <ipc/services.h>
+#include <ipc/input.h>
+#include <abi/ipc/interfaces.h>
+#include <loc.h>
+
+static const char *cmdname = "layout";
+
+/* Dispays help for layout in various levels */
+static void print_help()
+{
+	printf("Changes, lists or displays the current keyboard layout.\n");
+
+	printf(
+	    "Usage: %s\n"
+	    "\t%s list\tlists all layouts\n"
+	    "\t%s get\t displays currently set layout\n"
+	    "\t%s set <layout>\tchanges to the new layout\n",
+	    cmdname, cmdname, cmdname, cmdname);
+}
+
+/* lists all available kb layouts */
+static errno_t list_layout(void)
+{
+	/* TODO: replace this with the content of the folder containing the keymaps */
```

Comment:
I think rather the TODO should say we should query the server for available maps.

Reply:
When I implemented this function I thought about this and came to the conclusion that querying the server is unnecessary complicated with no benefit. It's an extra IPC call with no real advantage? I think it's easier if the server receives a path with the location of the keymap. Displaying a list (and translating that list) is a job done by the app